frontend/src/metabase/visualizations/lib/LineAreaBarRenderer.js

```diff
@@ -224,6 +224,64 @@ function getDimensionsAndGroupsAndUpdateSeriesDisplayNamesForStackedChart(
   return { dimension, groups };
 }
 
+// ASSERT(datas.length === 1)
+// ASSERT(datas[0].length > 0)
+function getDimensionsAndGroupsForWaterfallChart(props, originalDatas, warn) {
+  const datas = originalDatas.slice();
+  datas.push(datas[0].map(k => k.slice())); // negatives
+  datas.push(datas[0].map(k => k.slice())); // positives
+
+  /*
+  A waterfall chart is essentially a stacked bar chart.
+  It consists of the following (from the topmost):
+  - the "green bar", when the current value is greater than the previous one
+  - the "red bar", when the current value is smaller than the previous one
+  - the "invisible beam" supporting either the positive or negative bar
+
+  Note the green and red bars are mutually exclusive (i.e. if one has
+  a positive value, the other is zero). This is done so we need not have
+  conditional fill color.
+  */
+
+  const values = datas[0].map(d => d[1]);
+  const positives = values.map(v => (v > 0 ? v : 0));
+  const negatives = values.map(v => (v < 0 ? -v : 0));
+  const beams = [0];
+  for (let i = 1; i < values.length; ++i) {
+    beams[i] =
+      positives[i - 1] > 0
+        ? beams[i - 1] + positives[i - 1] - negatives[i]
```

Comment:
This logic breaks if it starts with a negative, but I think we knew that was a weird case that could be handled after the feature took shape.